blackboxopt/optimizers/staged/optimizer.py

```diff
@@ -46,20 +46,33 @@ def __init__(
         self.evaluation_uuid_to_iteration: Dict[str, int] = {}
         self.pending_configurations: Dict[str, EvaluationSpecification] = {}
 
-    def report_evaluation(self, evaluation: Evaluation) -> None:
-        super().report_evaluation(evaluation)
+    def report(self, evaluations: Union[Evaluation, Iterable[Evaluation]]) -> None:
+        super().report(evaluations)
 
-        evaluation_specification_id = evaluation.optimizer_info.get("id")
-        if evaluation_specification_id is None:
-            raise ValueError(
-                "Missing evaluation specification ID in optimizer info. Did you try to "
-                + "report an evaluation for a configuration which the optimizer did not"
-                " pick? This is not supported at the moment."
+        if isinstance(evaluations, Evaluation):
+            evaluations = [evaluations]
+
+        invalid_evaluations = []
+        for evaluation in evaluations:
+            if evaluation.optimizer_info.get("id") is None:
+                invalid_evaluations.append(evaluation)
+                continue
+
+            evaluation_specification_id = evaluation.optimizer_info.get("id")
+            self.pending_configurations.pop(str(evaluation_specification_id))
+            idx = self.evaluation_uuid_to_iteration.pop(
+                str(evaluation_specification_id)
+            )
+            self.iterations[idx].digest_evaluation(
+                evaluation_specification_id, evaluation
             )
 
-        self.pending_configurations.pop(str(evaluation_specification_id))
-        idx = self.evaluation_uuid_to_iteration.pop(str(evaluation_specification_id))
-        self.iterations[idx].digest_evaluation(evaluation_specification_id, evaluation)
+        if invalid_evaluations:
+            raise EvaluationsError(
+                f"{len(invalid_evaluations)} evaluation(s) got rejected because of a "
+                "missing an evaluation specification ID in the optimizer info.",
+                evaluations=invalid_evaluations,
+            )
```

Comment:
To combat that separation of where the error is detected and where it's reported, let's rename the variable to `evaluations_without_ids`, because otherwise one might happily add evaluations that are invalid for other reasons to the variable and get a mismatched error message.

Unrelated, the error message should read "[...] because of missing an evaluation [...]"